test/test_sum_product.py

```diff
@@ -86,21 +87,616 @@ def test_partition(inputs, dims, expected_num_components):
+    expected = sum_product(sum_op, prod_op, unrolled_factors, reduce_vars, plates)
+
+    return expected
+
+
+@pytest.mark.parametrize('vars1,vars2', [
+    (frozenset(),
+     frozenset({"sequences", "time", "x_prev", "x_curr", "tones", "y_prev", "y_curr"})),
+    (frozenset({"time", "x_prev", "x_curr", "tones", "y_prev", "y_curr"}),
+     frozenset({"sequences"})),
+    (frozenset({"sequences", "time", "x_prev", "x_curr", "y_prev", "y_curr"}),
+     frozenset()),
+])
+@pytest.mark.parametrize('sum_op,prod_op', [(ops.logaddexp, ops.add), (ops.add, ops.mul)])
+def test_partial_sum_product_hmm_example_0(sum_op, prod_op, vars1, vars2):
+    x_dim, y_dim = 2, 3
+    sequences, duration, tones = 2, 5, 4
+
+    probs_x = random_tensor(OrderedDict({}))
+    probs_y = random_tensor(OrderedDict({}))
+
+    x = random_tensor(OrderedDict({
+        "sequences": Bint[sequences],
+        "time": Bint[duration],
+        "x_prev": Bint[x_dim],
+        "x_curr": Bint[x_dim],
+    }))
+
+    y = random_tensor(OrderedDict({
+        "sequences": Bint[sequences],
+        "time": Bint[duration],
+        "tones": Bint[tones],
+        "y_prev": Bint[y_dim],
+        "y_curr": Bint[y_dim],
+    }))
+
+    factors = [probs_x, probs_y, x, y]
+    time = "time"
+    step = {"x_prev": "x_curr", "y_prev": "y_curr"}
+    plates = frozenset({"sequences", "tones"})
+    global_vars = frozenset()
+    local_vars = frozenset()
+
+    factors1 = modified_partial_sum_product(sum_op, prod_op, factors, vars1, plates, time, step)
+    factors2 = modified_partial_sum_product(sum_op, prod_op, factors1, vars2, plates, time, step)
+    actual = reduce(prod_op, factors2)
+
+    expected = _expected_hmm_example(sum_op, prod_op, factors, plates, global_vars, local_vars,
+                                     time, duration, step)
+
+    assert_close(actual, expected, atol=5e-4, rtol=5e-4)
+
+
+@pytest.mark.parametrize('vars1,vars2', [
+    (frozenset(),
+        frozenset({"sequences", "time", "x_prev", "x_curr", "tones", "y_curr"})),
+    (frozenset({"y_curr", "tones"}),
+        frozenset({"sequences", "time", "x_prev", "x_curr"})),
+    (frozenset({"time", "x_prev", "x_curr", "tones", "y_curr"}),
+        frozenset({"sequences"})),
+    (frozenset({"sequences", "time", "x_prev", "x_curr", "tones", "y_curr"}),
+        frozenset()),
+])
+@pytest.mark.parametrize('sum_op,prod_op', [(ops.logaddexp, ops.add), (ops.add, ops.mul)])
+def test_partial_sum_product_hmm_example_1(sum_op, prod_op, vars1, vars2):
+    x_dim, y_dim = 2, 3
+    sequences, duration, tones = 2, 5, 4
+
+    probs_x = random_tensor(OrderedDict({}))
+    probs_y = random_tensor(OrderedDict({}))
+
+    x = random_tensor(OrderedDict({
+        "sequences": Bint[sequences],
+        "time": Bint[duration],
+        "x_prev": Bint[x_dim],
+        "x_curr": Bint[x_dim],
+    }))
+
+    y = random_tensor(OrderedDict({
+        "sequences": Bint[sequences],
+        "time": Bint[duration],
+        "tones": Bint[tones],
```

Comment:
When are the plates `tones` and `sequences` being eliminated by `modified_partial_sum_product`? It should be before and after `time` respectively, rather than both before or both after.

Reply:
For different combinations of `vars1` and `vars2` it follows that order to eliminate plates. Second case is when `tones` removed before `time` and third case is `sequences` eliminated after `time`.